drv/transceivers-server/src/net_api.rs

```diff
@@ -0,0 +1,410 @@
+        }
+    }
+
+    /// Handles a single message from the host with supplementary data in `data`
+    ///
+    /// Returns a response and a `usize` indicating how much was written to the
+    /// `out` buffer.
+    fn handle_message(
+        &mut self,
+        msg: Message,
+        data: &[u8],
+        out: &mut [u8],
+    ) -> Result<(HostResponse, usize), Error> {
+        if msg.header.version != 1 {
+            return Err(Error::VersionMismatch);
+        }
+
+        match msg.body {
+            MessageBody::SpRequest(..)
+            | MessageBody::SpResponse(..)
+            | MessageBody::HostResponse(..) => {
+                return Err(Error::ProtocolError);
+            }
+            MessageBody::HostRequest(h) => {
+                self.handle_host_request(h, msg.modules, data, out)
+            }
+        }
+    }
+
+    fn handle_host_request(
+        &mut self,
+        h: HostRequest,
+        modules: ModuleId,
+        data: &[u8],
+        out: &mut [u8],
+    ) -> Result<(HostResponse, usize), Error> {
+        match h {
+            HostRequest::Reset => {
+                let mask = get_mask(modules)?;
+                // TODO: use a more correct error code
+                self.transceivers
+                    .set_reset(mask)
+                    .map_err(|_e| Error::ReadFailed)?;
+                userlib::hl::sleep_for(1);
+                self.transceivers
+                    .clear_reset(mask)
+                    .map_err(|_e| Error::ReadFailed)?;
+                Ok((HostResponse::Ack, 0))
+            }
+            HostRequest::Status => {
+                use drv_sidecar_front_io::Addr;
+                use zerocopy::{BigEndian, U16};
+
+                let fpga = match modules.fpga_id {
+                    0 => FpgaController::Left,
+                    1 => FpgaController::Right,
+                    i => return Err(Error::InvalidFpga(i)),
+                };
+                let fpga = self.transceivers.fpga(fpga);
+
+                // This is a bit awkward: the FPGA will get _every_ module's
+                // status (for the given FPGA), then we'll unpack to only the
+                // ones that we care about
+                let enable: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_CTRL_EN_H)
+                    .map_err(|_e| Error::ReadFailed)?;
+                let reset: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_CTRL_RESET_H)
+                    .map_err(|_e| Error::ReadFailed)?;
+                let lpmode: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_CTRL_LPMODE_H)
+                    .map_err(|_e| Error::ReadFailed)?;
+                let present: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_STATUS_PRESENT_H)
+                    .map_err(|_e| Error::ReadFailed)?;
+                let irq: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_STATUS_IRQ_H)
+                    .map_err(|_e| Error::ReadFailed)?;
```

Comment:
`transceivers` also has a `get_module_status` function that may be of use to you here. It will get all of these things in one larger read. Then you can go about masking them off below. https://github.com/oxidecomputer/hubris/blob/master/drv/sidecar-front-io/src/transceivers.rs#L241-L276

Reply:
I looked at that, but it's not quite what I want: `get_module_status` returns _logical_ port status, and here we want _physical_ port status (since the host side is doing the physical -> logical decoding itself).